### ISSUE TITLE
multistage builds leak linux semaphore arrays which prevent creating images or containers

### ISSUE THREAD

**Issue by jaloren** [moby/moby#34564]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

**Description**
I have a dockerfile that uses the multistage build feature. During the docker build, over 128 linux semaphore arrays are generated which exhausts the linux kernal limit. 

**Steps to reproduce the issue:**
1. create a docker file that performs a multistage build where the base image is centos 7.
2. run docker build.

**Describe the results you received:**
Docker build fails to create an image because the linux semaphore arrays are exhausted.

**Describe the results you expected:**
docker build successfully creates the image.

**Additional information you deem important (e.g. issue happens only occasionally):**

This is occurring with the device mapper storage layer. Here's the docker build file i used.

```
FROM private-registry:8082/centos-7 as build

RUN set -x \
    && yum install -y perl make automake gcc gmp-devel libffi zlib xz tar git gnupg 

FROM private-registry:8082/centos-7

COPY --from=build /root/ /tmp/root
```

I am using centos 7.3. Here's uname information:

```
Linux build-slave-9 3.10.0-514.21.2.el7.x86_64 #1 SMP Tue Jun 20 12:24:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:20:36 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:21:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 26
Server Version: 17.06.0-ce
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 9.473GB
 Data Space Total: 25.5GB
 Data Space Available: 16.02GB
 Metadata Space Used: 1.098MB
 Metadata Space Total: 264.2MB
 Metadata Space Available: 263.1MB
 Thin Pool Minimum Free Space: 2.55GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.21.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.639GiB
Name: build-slave
ID: DUDO:TN33:WPJY:DM5Z:TZXZ:F6JU:I24K:2A62:XO2G:GC7V:IQKL:W775
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by jaloren** [comment]
This might be resolved this PR is merged. 

https://github.com/docker/docker-ce/pull/200

**Comment by thaJeztah** [comment]
Yes, this looks like https://github.com/docker/docker-ce/pull/200 (and https://github.com/moby/moby/pull/33732), and should be resolved in Docker 17.06.2.

I'll go ahead and close, but let me know if you still encounter this after updating to that version
